### ISSUE TITLE
@InjectMocks ignores @Mock's name of generic types

### ISSUE THREAD

**Issue by Ghedeon** [mockito/mockito#1066]
When injecting two fields of the same type with `@InjectMocks`, the `@Mock`'s name is not taken into account.

```java
public class Foo {
    Bar<Integer> f1;
    Bar<Boolean> f2;

    public Foo(Bar<Integer> f1, Bar<Boolean> f2) {
        this.f1 = f1;
        this.f2 = f2;
    }
}
```

```java
interface Bar<T> {
}
```

```java
@RunWith(MockitoJUnitRunner.class)
public class FooTest {

    @Mock(name = "f1")
    Bar<Integer> f1;

    @Mock(name = "f2")
    Bar<Boolean> f2;

    @InjectMocks
    Foo foo;

    @Test
    public void test() {
        System.out.println(foo.f1);
        System.out.println(foo.f2);
    }
}
```
**Expected result**: f1 mock --> f1, and f2 mock --> f2.
**Actual result**: f1 mock is getting injected into both fields.

Version: mockito-core:2.7.22

**Comment by TimvdLippe** [comment]
Hm, not sure if we can cope with that. We are not aiming to be a full injection framework. If there is an easy fix, we can do that. Else we have to document this deficiency :(

**Comment by Ghedeon** [comment]
@TimvdLippe Totally understand your point, but what is the name attribute for if not for this? It's literally documented as expected behavior for `@InjectMocks`. It's a canonical example from @InjectMocks javadoc (https://static.javadoc.io/org.mockito/mockito-core/2.7.22/org/mockito/InjectMocks.html) and it's not working. It's a trivial use case, there is nothing tricky or hackish about this injection.

**Comment by TimvdLippe** [comment]
Oh right. Hm yes that should be easier to fix and we should do that. Do you have time to submit a PR?

**Comment by ChristianSchwarz** [comment]
Indeed it would be nice if Mockito would support construtor injection by parameter name! Constructor injection is different beast that field injection. For field injection the name of the target field can be used to match to the mock of the same name. For parameters this is not always possible, it depends on how the code was compiled  and on the java version that run the code:
 * till Java 8 there was [no API to retrieve the parameter name](https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Constructor.html), all parameters are named "arg0",arg1",...
 * in Java 8  [Executable.getParameters](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Executable.html#getParameters--) was introduced, that allow to retrieve the parameter name with limitations:

> Method Parameter Reflection (RFE: JDK-8004841): You can obtain the names of the formal parameters of any method or constructor with the method java.lang.reflect.Executable.getParameters. However, .class files do not store formal parameter names by default. To store formal parameter names in a particular .class file, and thus enable the Reflection API to retrieve formal parameter names, compile the source file with the _-parameters_ option of the javac compiler.


----
Long story short: It is possibe to implement construtor injection via parameter names. But your tests may break if they are not compiled or run with the same setup. 


This limitation should be documented. A workaround would be to name the mock "argN" to indicate the parameter index for injection, but this has many other downsides.

**Comment by msangel** [comment]
All proposed above have sense. And having `@Mock(name="argN")` is index  is better then nothing.

Found that funny, but I never saw situation when tests was compiled and run on different env.
>tests may break if they are not compiled or run with the same setup

**Comment by ubhide** [comment]
Just to keep this bug alive, I am definitely having this issue in v1.10.19

My workaround was to not use InjectMocks and create the object manually in the setup. Kinda messy but it gets around this issue.

**Comment by Meaaow** [comment]
I am also facing the same issue - #2041
I used reflection to create the object myself.

**Comment by tuxslayer** [comment]
Still relevant :(

**Comment by bubalehich** [comment]
Any updates?

**Comment by konikvranik** [comment]
Still not solved?

**Comment by TotoroDev-eqi** [comment]
Is it going to be fixed?

**Comment by CQ-Developer** [comment]
still not working T-T

**Comment by ChristianSchwarz** [comment]
The moste reliable and easy fix would be to specify the the arguments to inject in the `@InjectMock` annotation.

```
    @Mock(name = "f1")
    Bar<Integer> f1;

    @Mock(name = "f2")
    Bar<Boolean> f2;

    @InjectMocks(useMock=["f1","f2"])
    Foo foo;
```

With the parameter `useMock` the order  and count of the mocks to be passed to the constructor is clear. If the types or number of arguments doesn't match any constructor, mockito should simply throw. In case more than one constructor matches, an optional annotation argument can be introduced to describe the constructor signature targeted, like `argTypes=[Integer.class,Boolean.class]`. Mockito has then all information needed to find a matching constructor.

**Comment by CQ-Developer** [comment]
@ChristianSchwarz I'm fixed this bug, and tried to submit a pr, but got 403.
Can you give me permission?

**Comment by rfelgent** [comment]
still not working
